db/migrate/20210119082215_add_username_to_users.rb

```diff
@@ -0,0 +1,6 @@
+class AddUsernameToUsers < ActiveRecord::Migration[5.2]
+  def change
+    add_column :users, :username, :string, null: true
+    add_index :users, :username, unique: true, where: 'username IS NOT NULL'
```

Comment:
Is this `where` doing anything?
Can you show me the `show create table` ?

Reply:
The where clause will check to validate uniqueness for not-null username. I tested it works. ref: https://stackoverflow.com/a/48147333/5087746.

here is `show create table`
```sql
CREATE TABLE `users` (
  `id` bigint NOT NULL AUTO_INCREMENT,
  `uid` varchar(255) NOT NULL,
  `email` varchar(255) NOT NULL,
  `password_digest` varchar(255) NOT NULL,
  `role` varchar(255) NOT NULL DEFAULT 'member',
  `data` text,
  `level` int NOT NULL DEFAULT '0',
  `otp` tinyint(1) DEFAULT '0',
  `state` varchar(255) NOT NULL DEFAULT 'pending',
  `referral_id` bigint DEFAULT NULL,
  `created_at` datetime NOT NULL,
  `updated_at` datetime NOT NULL,
  `username` varchar(255) DEFAULT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `index_users_on_uid` (`uid`),
  UNIQUE KEY `index_users_on_email` (`email`),
  UNIQUE KEY `index_users_on_username` (`username`)
) ENGINE=InnoDB AUTO_INCREMENT=3 DEFAULT CHARSET=utf8
```